### ISSUE TITLE
Weather code is wildly off

### ISSUE THREAD

**Issue by krschacht** [open-meteo/open-meteo#812]
Right now in Austin, Texas the current weather code being returned is 95 (thunderstorms). This is wildly off. All of the other daily and current weather details I'm getting are accurate: it's a day filled with hot clear skies.

This is my first attempt to retrieve the weather_code, why is it contradicting the reality of the weather situation?

**Comment by krschacht** [comment]
As I investigate further, it also thinks there is growing cloud cover over Austin right now with forecasted 100% cloud cover in a couple hours. This is definitely wrong.

I've tried altering the lat & lon slightly in case there was just a quirk of the data, but it's reporting the same for really close towns (such as Rollingwood, Texas).

Here is the results from the hourly pull that I just did:

```
"hourly": {
"time": [
"2024-05-25T00:00",
"2024-05-25T01:00",
"2024-05-25T02:00",
"2024-05-25T03:00",
"2024-05-25T04:00",
"2024-05-25T05:00",
"2024-05-25T06:00",
"2024-05-25T07:00",
"2024-05-25T08:00",
"2024-05-25T09:00",
"2024-05-25T10:00",
"2024-05-25T11:00",
"2024-05-25T12:00",
"2024-05-25T13:00",
"2024-05-25T14:00",
"2024-05-25T15:00",
"2024-05-25T16:00",
"2024-05-25T17:00",
"2024-05-25T18:00",
"2024-05-25T19:00",
"2024-05-25T20:00",
"2024-05-25T21:00",
"2024-05-25T22:00",
"2024-05-25T23:00"
],
"weather_code": [
95,
0,
1,
3,
1,
3,
3,
3,
95,
95,
95,
95,
95,
95,
95,
95,
95,
95,
0,
0,
0,
0,
0,
0
],
"cloud_cover": [
1,
0,
45,
100,
31,
100,
100,
100,
100,
98,
99,
36,
33,
5,
0,
0,
2,
4,
1,
0,
0,
0,
0,
3
],
```

**Comment by krschacht** [comment]
Hey @patrick-zippenfenig, is this a known issue?

**Comment by patrick-zippenfenig** [comment]
Hi,
partly. Thunderstorm weather codes are calculated based in thresholds for lifted index and CAPE. I am considering to also include additional factors like a minimum of 30% cloud cover or minimum of 20% precipitation probability. Could you send the API URL you used to get this data? Did you set a timezone?

**Comment by krschacht** [comment]
Yes, I set the timezone. But I don't think it was a simple heuristic tweak because even the cloud cover prediction was way off. If you look at the data dump I pasted above, even the cloud cover was predicted to be 100% within a couple hours of me pulling it. And that was no where the case, it was a clear sky just partly cloudy. Every other weather service I checked was also predicting clear/partly cloudy skies that day with 0% chance of rain.

The API call was:
https://api.open-meteo.com/v1/forecast?past_days=1&forecast_days=1&latitude=30.26715&longitude=-97.74306&timezone=America%2FChicago&temperature_unit=fahrenheit&wind_speed_unit=mph&precipitation_unit=inch&current=temperature_2m%2Capparent_temperature%2Cprecipitation%2Crain%2Cshowers%2Csnowfall%2Ccloud_cover&hourly=temperature_2m%2Capparent_temperature%2Cprecipitation_probability%2Cprecipitation%2Crain%2Cshowers%2Csnowfall%2Csnow_depth%2Ccloud_cover&daily=temperature_2m_max%2Ctemperature_2m_min%2Capparent_temperature_max%2Capparent_temperature_min%2Cprecipitation_sum%2Crain_sum%2Cshowers_sum%2Csnowfall_sum%2Cprecipitation_probability_max

I can't reconstruct the full output right now but when I was diagnosing that particular morning, I pulled the hourly results and that's what I pasted above. It was the hourly results that suggested to me: something seems really off with whatever the source of truth is for this particular moment.

**Comment by krschacht** [comment]
@patrick-zippenfenig I came back to this API to add some additional improvements to my app, and once again the forecast is off. It's reporting thunderstorms tomorrow (8/19) while other web services are reporting clear skies with 1% chance of rain:

This is my API call:

https://api.open-meteo.com/v1/forecast?daily=weather_code%2Ctemperature_2m_max%2Ctemperature_2m_min%2Capparent_temperature_max%2Capparent_temperature_min%2Cprecipitation_sum%2Crain_sum%2Cshowers_sum%2Csnowfall_sum%2Cprecipitation_probability_max&forecast_days=2&latitude=30.26715&longitude=-97.74306&precipitation_unit=inch&temperature_unit=fahrenheit&timezone=America%2FChicago&wind_speed_unit=mph

**Comment by patrick-zippenfenig** [comment]
Hi,

it looks like the following model updates removed the thunderstorm pictogram.

There are a couple of issues here:
- The pictogram always shows the worst condition within 24 hours. Regardless if it is in the afternoon or at 3 a.m.
- Thunderstorms are approximated from the HRRR weather model using CAPE and Lifted Index. Here are the [thresholds](https://github.com/open-meteo/open-meteo/blob/9bc8c8d5571a108ca4bb7db4128386a4506ad141/Sources/App/Helper/WeatherCode.swift#L44). As mentioned before, adding cloud cover and precipitation probability could improve it, but I had no time to validate this approach.

If you would like to experiment, you can retrieve CAPE, lifted index and other data directly and define your own thresholds.

**Comment by krschacht** [comment]
Got it, I appreciate the further input. I was hoping it may just be a quick fix, but I get that you need to do some back testing to validate. I’ll probably not chase this down any further for now. Weather is just one of many information sources I’m pulling into a voice AI assistant so I’ll just “fix” for now by ignoring the weather code.

**Comment by phennessey** [comment]
I have this exact issue. Icon is showing up as 95 (thunderstorms) consistently for the last 48 hours despite zero precipitation and no storm activity across the entire state in which I live.

Look at the last 2 weeks: It behaved normally but then over the last 3 days it starts to erratically jump to 95 (thunderstorms) despite no real reason to do so, with a striking regularity only recently. @patrick-zippenfenig this looks like a bug in the approximation, but I'm not sure:

<img width="1410" height="393" alt="Image" src="https://github.com/user-attachments/assets/59dc3030-b5b9-45ac-bb53-2451e81b6769" />
<a href="https://api.open-meteo.com/v1/forecast?latitude=39.1180253&longitude=-84.4907521&hourly=temperature_2m,weather_code,precipitation&past_days=14&forecast_days=1" />open-meteo API call</a><br /><br />

Also, take a look at this forecast, which exhibits the same erratic icon behavior:
<img width="1553" height="400" alt="Image" src="https://github.com/user-attachments/assets/41aee60f-9bcb-4292-8516-c7e3e91b92e8" />
<a href="https://api.open-meteo.com/v1/forecast?latitude=39.1180253&longitude=-84.4907521&hourly=temperature_2m,weather_code,precipitation" />open-meteo API call</a>

**Comment by strumswell** [comment]
Yes, I have also noticed this with users in the US (Minneapolis) and cross-checked with data coming from weather.gov which behaves correctly by not reporting constant thunderstorms. I looked a little into the code and I think this is due to `convectivePrecipitation` being used in `calculateThunderstormProbability` which might be nil or zero-filled in some models. I thought about opening up a PR where I pull in precipitation but after some reading I noticed that this seems to be the wrong approach. The linked COSMO/ DWD PDFs in the comment seem to suggest that some kind of precipitation is necessary though for a thunderstorm WMO code. I wonder what the team thinks about this.

**Comment by phennessey** [comment]
The culprit might be this block, which runs before any precipitation check:
```
if let cape, cape >= 3000 {
    if let liftedIndex {
        if liftedIndex <= -5 {
            return .thunderstormSlightOrModerate
        }
    } else {
        return .thunderstormSlightOrModerate
    }
}
```

Claude shared this response:
> If `liftedIndex` is `nil` (i.e., that field isn't populated for whatever model/location you're pulling from), the code unconditionally falls back to `.thunderstormSlightOrModerate` the moment CAPE crosses 3000 J/kg — with zero regard for actual precipitation. Given your data shows plenty of hours with `precipitation: 0` and `weather_code: 95` side by side, that's almost certainly what's happening: Cincinnati in July routinely produces high CAPE (hot, humid, unstable air) even on dry days, and if `liftedIndex` isn't coming through from your data source, every one of those high-CAPE hours gets short-circuited into a thunderstorm code regardless of whether rain is actually falling. The fix is to stop treating a missing `liftedIndex` as an automatic "yes" — either require some minimum precipitation/convective precipitation before returning a thunderstorm code, or treat a nil `liftedIndex` as inconclusive (skip to the precipitation-based logic) rather than a green light.

**Comment by patrick-zippenfenig** [comment]
> Yes, I have also noticed this with users in the US (Minneapolis) and cross-checked with data coming from weather.gov which behaves correctly by not reporting constant thunderstorms. I looked a little into the code and I think this is due to `convectivePrecipitation` being used in `calculateThunderstormProbability` which might be nil or zero-filled in some models. I thought about opening up a PR where I pull in precipitation but after some reading I noticed that this seems to be the wrong approach. The linked COSMO/ DWD PDFs in the comment seem to suggest that some kind of precipitation is necessary though for a thunderstorm WMO code. I wonder what the team thinks about this.

Good point. The thunderstorm probability calculation was updated just 2 weeks ago and yes, some models do not provide convective precipitation. I am working on a change to better consider precipitation and additionally reduce thunderstorm probability at low cloud cover.

**Comment by phennessey** [comment]
Looks like the pull request improved this behavior a bit. Still a few code-95 spikes, but definitely better overall.

<img width="1158" height="383" alt="Image" src="https://github.com/user-attachments/assets/e2886cdd-901d-498c-b1e5-69bdde45a19d" />
<img width="1153" height="381" alt="Image" src="https://github.com/user-attachments/assets/887d2bce-819e-4394-b67d-2bca43d3c09b" />

**Comment by patrick-zippenfenig** [comment]
I slightly adjusted the thresholds to show fewer thunderstorms if no precipitation is set. Of course it is only a numerical approximation so there will always be cases showing too many or few thunderstorms. I think the current thresholds are a good middle ground

**Comment by strumswell** [comment]
Thank you for taking care of this, Patrick. You guys are doing amazing work.
